oeps/oep-0019-bp-developer-documentation.rst

```diff
@@ -0,0 +1,450 @@
+       * *What are the terms and definitions of this component's ubiquitous language?*
+       * *How can I navigate the documentation within and related to this component?*
+   * - **Rationale**
+     - The future evolution and development of a component remains within the intended
+       boundaries of the component's responsibility. Otherwise, it invites unintended complexity
+       and incongruity.
+   * - **Location in GitHub**
+     - Since GitHub automatically renders the contents of the repo's top-level README.rst file
+       on the repo's home page, README files will be located at the top-level of their corresponding
+       repo/app/folder.
+   * - **Maintainence**
+     - The development team that owns the component should keep its README updated.
+
+How-Tos
+-------
+
+.. list-table::
+   :widths: 15 75 
+
+   * - **Description**
+     - How-To documents provide relatively long-term step-by-step instructions in using a 
+       feature/component/etc. For instructions that are expected to be ephemeral and short-lived,
+       other alternatives such as `Etsy's FYIs`_ can be explored.
+   * - **Answers the questions**
+     - 
+       * *How do I install/enable/use/test/etc this component?*
+   * - **Rationale**
+     - To provide concise recipe-style instructions to users of the component so they can
+       get started quickly without needing to understand the technical nuances and code of
+       the component.
+   * - **Location in GitHub**
+     - A how_tos_ folder will be created within a **docs** folder in the repo/app/folder. 
+       The goal is to keep the How-Tos in synch with the code so that (1) developers running
+       older versions of the code have the appropriately versioned information and (2) developers
+       updating the component have easy access to update the corresponding How-Tos.
+   * - **Maintainence**
+     - The developer and code reviewer that updates the component will be responsible for updating
+       the relevant How-Tos in that repo (preferably in the same PR as the code changes).
+
+.. _how_tos: https://github.com/edx/edx-platform/tree/master/openedx/core/djangoapps/oauth_dispatch/docs/how_tos
+.. _`Etsy's FYIs`: https://codeascraft.com/2018/10/10/etsys-experiment-with-immutable-documentation/
+
+APIs
+----
+
+.. list-table::
+   :widths: 15 75 
+
+   * - **Description**
+     - API documentation provide information on interfaces for integrating with or enhancing the
+       platform.
+   * - **Answers the questions**
+     - 
+       * *What are the public and supported APIs for this component?*
+       * *What are the request/response/authentication/etc protocol for this component?*
+   * - **Rationale**
+     - Developers that use and integrate with the component have a clearly defined and consistenly
+       documented interface.
+   * - **Location and tools**
+     - The documentation of APIs should be automatically generated and extracted from the code
```

Comment:
I think it's a narrow opinion that all API documentation should be automatically generated. What if there are developers who are not familiar with the mechanism to auto generate documentations? What if there are API performance characteristics we want to express in this documentation that cannot be written in auto generated docs? I think it's better to widen this opinion to provide guidance on what to do for manually written API docs. I am wondering why automatically generated documentation should be enough for API?

Reply:
I believe auto-generated docs include the docstrings that code authors put into the code, no?

Reply:
@nedbat Yes. I believe so. However, auto-generated documentation represent an additional overhead for reading them. I am wondering the cost of complying with auto generated doc format is worth it for the standardization of API docs.

Reply:
@schenedx Is there a scalable and maintainable alternative that you recommend?  Prior to this, we tried having a super-small edX Documentation team write and update API docs.  However, that had several issues: (1) the document and code easily diverged and were continually out-of-date, (2) we needed to invest in specialized resources in the documentation team, and (3) API users need more than just the documentation - they need a tool to try and test out the APIs, which tools like Swagger support.